src/pyggdrasil/tree_inference/_logprob.py

```diff
@@ -0,0 +1,143 @@
+
+    Returns:
+        likelihood of the cell / mutation - see Equation 13
+        log (P(D_{ij} | A(T)_{i˜sigma_j}) )
+            i / axis 0 has n dimensions sum over mutation nodes (n)
+                - each mutation
+            j / axis 1 has m dimensions sum over cells (m)
+                - each cell
+            k / axis 2 has n+1 dimensions sum over nodes (n+1)
+                - attachment to mutation and root
+    """
+
+    # A(T) - get ancestor matrix
+    ancestor_mat = tr._get_ancestor_matrix(tree.tree_topology)
+    # get mutation likelihood
+    mutation_likelihood = _mutation_likelihood(mutation_mat, ancestor_mat, theta)
+    # get log of it
+    log_mutation_likelihood = jnp.log(mutation_likelihood)
+
+    return log_mutation_likelihood
+
+
+def _mutation_likelihood(
+    mutation_matrix: MutationMatrix,
+    ancestor_matrix: AncestorMatrix,
+    theta: tuple[float, float],
+) -> Mutation_Likelihood:
+    """Returns the mutation likelihood given the data and expected mutation matrix.
+
+    Currently, only implements non-homozygous mutations.
+
+    Args:
+        mutation_matrix: mutation matrix
+        ancestor_matrix: ancestor matrix
+    Returns:
+        mutation likelihood - P(D_{ij} | A(T)_{i˜sigma_j})
+            i / axis 0 has n dimensions sum over mutation nodes (n)
+                - each mutation
+            j / axis 1 has m dimensions sum over cells (m)
+                - each cell
+            k / axis 2 has n+1 dimensions sum over nodes (n+1)
+                - attachment to mutation and root
+    Note:
+        let k = sigma_j
+    """
+
+    # m = number of cells, n-1 = number of mutations, ex root
+    n, m = mutation_matrix.shape
+    alpha, beta = theta
+
+    # TODO: Triple-check if this is the correct @ Pawel @ Gordon
+
+    # truncate ancestor matrix
+    ancestor_matrix = ancestor_matrix[:-1]
+
+    # repeat the ancestor matrix  - tensor of dimensions (n, m, n+1)
+    ancestor_tensor = jnp.repeat(ancestor_matrix[:, jnp.newaxis, :], m, axis=1)
+
+    # repeat the mutation matrix  - tensor of dimensions (n, m, n+1)
+    mutation_tensor = jnp.repeat(mutation_matrix[:, :, jnp.newaxis], n + 1, axis=2)
+
+    # compute the likelihood
+    mutation_likelihood = jnp.zeros((n, m, n + 1))
+    # P(D_{ij} = 0| E_{ik} = 0) = 1- alpha
+    mask = (mutation_tensor == 0) & (ancestor_tensor == 0)
+    mutation_likelihood = jnp.where(mask, 1 - alpha, mutation_likelihood)
+    # P(D_{ij} = 0| E_{ik} = 1) = beta
+    mask = (mutation_tensor == 0) & (ancestor_tensor == 1)
+    mutation_likelihood = jnp.where(mask, beta, mutation_likelihood)
+    # P(D_{ij} = 1| E_{ik} = 0) = alpha
+    mask = (mutation_tensor == 1) & (ancestor_tensor == 0)
+    mutation_likelihood = jnp.where(mask, alpha, mutation_likelihood)
+    # P(D_{ij} = 1| E_{ik} = 1) = 1 - beta
+    mask = (mutation_tensor == 1) & (ancestor_tensor == 1)
+    mutation_likelihood = jnp.where(mask, 1 - beta, mutation_likelihood)
```

Comment:
Does this look right? This is how I understand the math of $P(D_{ij} | A(T)_{i \sigma_j})$. It works as expected, but my intuition of what the operation represents has left me in the 3rd dimension.  These are my workings: 
[SCITE_mutation_likelihood.pdf](https://github.com/cbg-ethz/PYggdrasil/files/11289383/SCITE_mutation_likelihood.pdf)

Reply:
It looks very good to me! :slightly_smiling_face: 
(But I'm not 100% sure, so it's good that you confirmed on simple examples that it works as expected).

I think you have a tensor which dimensions $ijk$ represent $P(D_ij | A(T)_{ik}$. With this you're good to go with summation (`logsumexp` and `einsum`) as you wrote above.

By the way, a nice trick to do here: if you have the mutation profile of a *single* cell (so that $m=1$ and $j=0$ is the only value index $j$ can take), you have a two-dimensional array which is easier to understand. Then, if you "clone" the cell into $n$ cells the loglikelihood should multiply by a factor of $n$.

Reply:
Oh yes - of course. Just worked through in 2D. Got the intuition back :) Will write a dynamic 2D test and submit for your review if it passes this.